README.txt
-----------------------------------------------------------------------------------
Project: OpenGTS - Open GPS Tracking System
URL    : http://www.opengts.org
File   : README.txt
-----------------------------------------------------------------------------------

Notes:
 - Support for the CelltracGTS/Server requires downloading/installing the library
   support for the "ctrac.war" servlet. Please see the following for more information:
      http://www.geotelematic.com/CelltracGTS
 - Please refer to the included "OpenGTS_Config.pdf" file for all installation and 
   configuration information for the OpenGTS system.
 - When updating to a newer version of OpenGTS, it is highly recommended that the 
   following command be run to update the database with the latest table/column 
   changes:
     $  bin/dbAdmin.pl -tables=ca
   Or, run the following if alternate key fields also need to be updated:
     $  bin/dbAdmin.pl -tables=cak
 - On Windows 7, Internet Explorer 8, if you receive the error 
      Error [jsMapInit:[object error]
   then you may need to enable "Active Scripting" in the security tab.
 - A listing of the database tables and fields can be found in the file "SCHEMA.txt"
   in the OpenGTS installation diretory.
 - Documentation is available online at the following link:
     http://www.opengts.org/documentation.html
 - This open-source version of OpenGTS includes support for the following GPS 
   tracking devices:
     - Aspicore GSM Tracker (Nokia, Samsung, Sony Ericsson phones).
       [http://www.aspicore.com]
     - Astra Telematics AT240, AT110, AT210.
       [http://www.gps-telematics.co.uk]
     - Sanav GC-101, MT-101, and CT-24 Personal Tracker (HTTP-based protocol).
       Sanav GX-101 Vehicle Tracker (HTTP-based protocol).
       [http://www.sanav.com]
     - V-Sun 3338 Personal Tracker.
     - CelltracGTS/Free Android application
     - CelltracGTS/Pro Android application 
     - Most TK102/TK103 devices.
     - ZhongShan SIPGEAR Technology Co, Ltd.
     - TAIP (Trimble ASCII Interface Protocol).
     - TrackStick GPS data logger.
     - "GPSMapper" capable phones.
     - "NetGPS" capable devices.

Please let us know if you find any issues with this release.

-----------------------------------------------------------------------------------

Visit the following link for additional downloadable documentation and add-ons for 
OpenGTS:
  http://www.opengts.org/documentation.html
  http://www.geotelematic.com/documentation.html

-----------------------------------------------------------------------------------

This package may contain one or more of the following source modules which contain 
copyrights from their respective authors.  Please review these source modules for 
their copyright and license information:
   war/track/js/mapstraction/mapstraction.js
   war/track/js/mapstraction/mapstraction-geocode.js
   war/track/js/mapstraction/labeledmarker.js
   war/track/js/sorttable/sorttable.js

-----------------------------------------------------------------------------------

Contributors:
  See the following link for a list of contributors:
    http://www.opengts.org/info.html

-----------------------------------------------------------------------------------

Sample data:
Additional documentation for installing sample 'demo' data into the database can be
found in the "README.txt" file in the "sampleData" directory at
    sampleData/README.txt

-----------------------------------------------------------------------------------

Sanav GC-101 GPS Tracking Device Support:
Additional documentation for installing and configuring the GC-101 server within
OpenGTS can be found in the "README.txt" file in the "gc101" source directory at
   src/org/opengts/war/gc101/README.txt

See the following link for manufacturer's product information:
   http://www.sanav.com

-----------------------------------------------------------------------------------

TAIP (Trimble ASCII Interface Protocol) Support:
Support for TAIP (Trimble ASCII Interface Protocol) is included in this release.  
This server uses the raw-socket mode device communication server based on the example
'template' server.  The "taip.jar" server jar file can be built with the command:
   > ant taip
And can be started in the same manner that other servers are started using the
"runserver.pl" command as follows:
   > $GTS_HOME/bin/runserver.pl -s taip

-----------------------------------------------------------------------------------

Updating existing Geozone table entries 
(only required when upgrading from versions previous to v1.9.3):

If you have existing Geozone table entries, they will need to be updated to include
Bounding-Box information used to optimize Geozone lookups.  This update process is
done automatically with the following steps:

1) Update the new Geozone bounding-box table fields:
     > bin/dbAdmin.pl -tables=ca

2) For each account having Geozone entries, run the following command (replace "<accountID>" 
   with the appropriate account-IDs for each account which has Geozone entries):
     > bin/admin.pl Geozone -account=<accountID> -list -update

The update process should now be complete.

-----------------------------------------------------------------------------------

Runtime config file property key references
(for *.conf files, such as 'default.conf', 'common.conf', 'webapp.conf, ...):
- The prefixing "#" is a comment character, indicating that the following text is for
  informational purposes only, and should not be parsed.  Comments must be placed on
  a separate line, and must not be specified at the end of a property specification.
- Previously defined property keys may be referenced by placing them in ${...} brackets,
  such as ${log.name}.  Environment variables may also be referenced in this manner, as in 
  ${GTS_HOME}, or ${CATALINA_HOME}.
- A default value may be specified for referenced property keys which have not been defined.   
  For instance ${OUTPUT_DIR=/tmp} would resolve to "/tmp", if the property OUTPUT_DIR is not 
  defined (as an environment variable, or otherwise), or will resolve to the value specified 
  on the OUTPUT_DIR property (or environment variable) if it is defined.
- Recursive propery key references may also be specified.  For instance, the specification
  ${THIS_DIR=${THAT_DIR=/tmp}} will first attempt to resolve the value for "THIS_DIR", if not
  found, then value of THAT_DIR will be returned.  If THAT_DIR is not defined, then finally
  the specification will resolve to "/tmp".
- Property reference specifications for which the property is not defined, and there is no
  specified default value will resolve to the literal "${var}" string.  That is, if "var"
  is not defined, then ${var} will resolve to the literal string "${var}".
- Property reference specifications are only resolved at the time a property value is 
  requested, not at the time the property key=value line is parsed (ie. lazy resolution).  
  For instance, in the property specification below, "var_a" will resolve to "test", even 
  though "var_b" is not defined until after "var_a":
    var_a=${var_b}
    var_b=test
  This means that if "var_b" should ever change, the resolved value of "var_a" will also
  change accordingly.  [ie. "var_a" will only be resolved at the time a call to an RTConfig
  property value retrieval method is called - such as RTConfig.getString("var_a")].
  Write-only (assignable) property keys (ie. %log, %include) are the only exception to this 
  rule.  Since their assigned behavior is executed at the time they are parsed ('%log' prints
  a log message, and '%include' includes a file), write-only property keys must be assigned
  values that can be fully resolved at the time their values are assigned.
- Property keys may be re-assigned.  Property keys take on their last assigned value.  For
  instance, the following will display "Hello", then display "World":
    test=Hello
    %log=${test}
    test=World
    %log=${test}
- Property key assignments can be conditional if the property key has not already been
  assigned a value by placing a "?" after the key specification.  For instance, in the 
  property assignment "KEY?=VALUE", the property "KEY" will only be assigned "VALUE" if 
  "KEY" has not already been assigned a value previously.
    
The following are reserved read-only constant property keys:
   %version        returns the current version, as in "2.3.4"
   %contextName    returns the context 'name' (name of the servlet context, or main class name)
   %contextPath    returns the context 'path' (path to the servlet context, or $GTS_HOME)
   %configURL      returns the URL for the loaded config file
   %hostName       returns the current host-name
   %hostIP         returns the current mail host IP address

The following are reserved write-only (assignable) constant property keys:
   %log            displays the specified value to the log output
   %include        includes the config at the value URL (URL must exist, error otherwise)
   %include?       includes the config at the value URL (URL may exist, otherwise ignored)
   (The included URL protocol must be one of "file" or "http".)
   
The following are available within a PrivateLabel session context:
   session.name    The PrivateLabel Domain name
   session.locale  The PrivateLabel Domain locale

-----------------------------------------------------------------------------------

All trade names listed above are trademarks of their respective companies.
OpenGTS and GTS Enterprise are not affiliated with any of the listed companies.

-----------------------------------------------------------------------------------

Contact Info:
Please feel free to contact us regarding questions on this package.

Thanks,
Martin D. Flynn
devstaff@opengts.org


CHANGELOG.txt
-----------------------------------------------------------------------------------
GTS Changelog
Copyright 2007-2016 GeoTelematic Solutions, Inc.
-----------------------------------------------------------------------------------

Version 2.6.1 [2016/01/17]
- NEW: Added additional status codes.
- NEW: Added additional fields to Device table.
- NEW: Added option to "checkInstall" to display list of modified files.
- NEW: Added "useSSL" option to Google reverse-geocoding configuration.
- NEW: Added support for OpenCageData reverse-geocoding.
- NEW: Map pushpin replay can be now configured to skip over nearby pushpins.
- NEW: Added "tripElapse" column to Driving/Stopped report
- NEW: Added support for fuel-tank #2 fuel level profiling
- NEW: Added additional geozone load/import options
- UPD: "startupInit.Device.OpenDMTPFieldInfo=true" needed to for OpenDMTP fields.
- FIX: Fixed index map link display for Event Detail (by Group) report
- DCS: astra] Added missing "estimateOdometer" support.
- DCS: tk10x] Added support for configuring hearbeat ACK.
- DCS: tk10x] Added "speedConversionFactor" property for custom speed units

Version 2.6.0 [2015/09/24]
- NEW: Added additional status codes.
- NEW: Requires at least Java JDK 7 to utilize additional features.
- NEW: Added "engineTemp", "oilCoolerInTemp", "oilCoolerOutTemp" to EventData.
- NEW: Added additional Device table columns.
- NEW: Updated checkInstall to report non-GTS user owned files in $CATALINA_HOME
- NEW: Geozone now supports setting custom status codes for Arrive/Depart.
- UPD: Status codes STATUS_DOOR_[OPEN|CLOSE]_* moved to new value.
- UPD: Status codes STATUS_SEATBELT_[ON|OFF]_* moved to new value.
- UPD: Added additional STATUS_INPUT_[ON|OFF]_* codes.
- UPD: "dcservers.dtd" updated with latest XML config (thanks to Andrey Troitskiy)
- UPD: CheckInstall now displays missing optional jars as warnings (not errors).
- UPD: CheckInstall now checks for non-GTS user owned files (Java 7+ only)
- UPD: Removed weather/cloud later from Google Maps (no longer supported by Google)
- DCS: astra] DCS updated with new protocol features.
- DCS: tk10x] DCS updated with new protocol features.

Version 2.5.9 [2015/05/15]
- NEW: Added additional status codes.
- NEW: Added "vehicleYear" to Device.java table module.
- NEW: Added "dutyStatus" to Driver.java table module.
- NEW: Added additional fields to EventData.java table module.
- NEW: Added separate city/state/zip/etc field parsing to Google reverse-geocoder
- NEW: ServerSocketThread "maxClientPoolSize" now also applied to UDP listeners
- NEW: Added "googleLink" support to "index" report columns
- NEW: Added additional redirect support to FileTools.java "copyFile(URL...")
- UPD: Nmea0183 module updated to handle additional custom record types.
- UPD: Added support for associating a pushpin with a Geozone on the Device map.
- FIX: "SeverSocketThread.java" now handles native thread creation error condition.
- FIX: Fixed Account.java "gtsDiskUsage" free space display.
- FIX: Fixed swapped badge#/license# display on Driver Admin list page.

Version 2.5.8 [2015/02/12]
- NEW: Initial support for downloading/installing CelltracGTS/Server plug-in.
- NEW: Added additional status codes.
- NEW: Added "places" support to Google map-provider.
- NEW: Initial support for handling non-linear analog sensors.
- NEW: Added support for setting DNS caching timeout.
- UPD: "EventData.allowInnoDBCountWithWhere" property defaults to "true".
- UPD: GeocodeProvider now checks property overrides in BasicPrivateLabelLoader.
- UPD: Motion Report "idle" definition changed to 0-speed during trip.
- DCS: Fixed Sanav CellTower information parsing (reorder CID/LAC).

Version 2.5.7 [2014/10/07]
- NEW: Added additional status codes.
- NEW: Initial support for DeviceGroup notify email address (incomplete).
- NEW: Added "useResultAddress" property to "nominatim" reverse-geocode provider.
- NEW: Added DATA_VEHICLE_MAKE, DATA_VEHICLE_MODEL for report FieldLayout.
- NEW: Added "fleetSortByField" property to MotionReport.
- FIX: Fixed odometer offset used by MotionReport.
- FIX: Fixed ignition-hour accoumulation in "Device.java".
- UPD: Check for "Timezones" tag within "Domain" in private.xml
- UPD: Changes made to support compiling under Java-8.
- UPD: Added comments to "org/opengts/war/tools/BufferedHttpServletResponse.java"
       to indicate the change required in this module to support compiling with
       Tomcat-8 (changes required due to API differences in Servlet Spec 3.1).
- DCS: tk10x] Added support for using the last valid GPS location.
- DCS: tk10x] Added support for VJoy device.
- DCS: tk10x] Cell-tower data parsing for tk103_2 (thanks to Franjieh El Khoury).

Version 2.5.6 [2014/06/29]
- NEW: Enabled "Driver Message Detail" report to display messages returned from
       from CelltracGTS/Pro application.
- NEW: Added EventData report columns "ptoDistance", "workHours", "workDistance".
- NEW: Added EventData columns FLD_ptoDistanceKM, FLD_workDistanceKM
- NEW: Added initial support for FuelLevelProfile.
- NEW: Initial implementation for active/inactive Geozone checks.
- FIX: Fixed battery-level percentage display on some reports.
- FIX: Fixed NPE that would occasionally occur when displaying "Near By" Geozones.

Version 2.5.5 [2014/05/05]
- NEW: Added additional status codes.
- NEW: Added new fields to EventData table
- NEW: Added support for fuel-cost to Account/Device table and reports.
- DCS: gprmc] Added battery-temperature support.
- DCS: gprmc] Automatically adjust non-decimal battery level value to a percent.

Version 2.5.4 [2014/04/02]
- NEW: Added additional status codes.
- NEW: Added "allowWebService" column to Account table.
- NEW: Added "insuranceExpire" column to Device table.
- UPD: Additional changes to support table management of indexes in MS SQL Server.
- UPD: Update CheckInstall to list currently loaded config files.
- DCS: gc101] Fix packets sent via POST
- DCS: gprmc] Fix packets sent via POST

src/org/opengts/Version.java
// ----------------------------------------------------------------------------
// Copyright 2007-2016, GeoTelematic Solutions, Inc.
// All rights reserved
// ----------------------------------------------------------------------------
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ----------------------------------------------------------------------------
// Change History:
//  2007/01/25 v1.2.0  [Initial release]
//  2007/02/08 v1.2.1,  2007/02/11 v1.3.0,  2007/02/16 v1.3.1,  2007/02/18 v1.3.2
//  2007/02/26 v1.3.3,  2007/02/28 v1.3.4,  2007/03/11 v1.3.5,  2007/03/16 v1.3.6,
//  2007/03/25 v1.3.7,  2007/03/30 v1.4.0,  2007/04/15 v1.4.1,  2007/04/17 v1.4.2,
//  2007/05/06 v1.4.3,  2007/05/20 v1.4.4,  2007/05/25 v1.4.5,  2007/06/03 v1.4.6,
//  2007/06/13 v1.4.7,  2007/06/14 v1.4.8,  2007/06/30 v1.5.0,  2007/07/14 v1.5.1,
//  2007/07/27 v1.5.2,  2007/08/09 v1.5.3,  2007/09/16 v1.6.0,  2007/11/28 v1.6.1,
//  2007/12/13 v1.6.2,  2008/01/10 v1.6.3,  2008/02/04 v1.6.4,  2008/02/07 v1.6.5,
//  2008/02/11 v1.6.6,  2008/02/17 v1.6.7,  2008/02/21 v1.6.8,  2008/02/27 v1.6.9,
//  2008/03/12 v1.6.10, 2008/03/28 v1.7.0,  2008/04/11 v1.7.1,  2008/05/14 v1.8.0,
//  2008/05/20 v1.8.1,  2008/05/22 v1.8.2,  2008/06/20 v1.8.3,  2008/07/08 v1.8.4,
//  2008/07/21 v1.9.0,  2008/07/27 v1.9.1,  2008/08/08 v1.9.2,  2008/08/15 v1.9.3,
//  2008/08/17 v1.9.4,  2008/08/20 v1.9.5,  2008/08/24 v1.9.6,  2008/09/01 v1.9.7,
//  2008/09/12 v1.9.8,  2008/09/19 v2.0.1,  2008/10/16 v2.0.2,  2008/12/01 v2.0.3,
//  2009/01/01 v2.0.4,  2009/02/01 v2.0.5,  2009/02/20 v2.0.6,  2009/04/02 v2.0.7,
//  2009/05/01 v2.0.8,  2009/05/24 v2.0.9,  2009/05/27 v2.1.0,  2009/06/01 v2.1.1,
//  2009/07/01 v2.1.2,  2009/08/09 v2.1.3,  2009/08/23 v2.1.4,  2009/08/25 v2.1.4.1,
//  2009/09/23 v2.1.5,  2009/10/05 v2.1.6,  2009/11/01 v2.1.7,  2009/11/10 v2.1.8,
//  2009/12/16 v2.1.9,  2010/01/29 v2.2.0,  2010/04/11 v2.2.1,  2010/04/25 v2.2.2,
//  2010/05/24 v2.2.3,  2010/06/17 v2.2.4,  2010/07/04 v2.2.5,  2010/07/18 v2.2.6,
//  2010/09/10 v2.2.7,  2010/10/25 v2.2.8,  2010/11/29 v2.2.9,  2011/01/28 v2.3.0,
//  2011/03/08 v2.3.1,  2011/04/01 v2.3.2,  2011/05/15 v2.3.3,  2011/06/16 v2.3.4,
//  2011/07/01 v2.3.5,  2011/07/15 v2.3.6,  2011/08/21 v2.3.7,  2011/10/10 v2.3.8,
//  2011/12/06 v2.3.9,  2012/02/03 v2.4.0,  2012/04/20 v2.4.1,  2012/05/27 v2.4.2,
//  2012/06/29 v2.4.3,  2012/08/01 v2.4.4,  2012/09/02 v2.4.5,  2012/11/02 v2.4.6,
//  2012/12/27 v2.4.7,  2013/03/01 v2.4.8,  2013/04/08 v2.4.9,  2013/06/05 v2.5.0,
//  2013/08/06 v2.5.1,  2013/09/30 v2.5.2,  2014/01/01 v2.5.3,  2014/03/03 v2.5.4,
//  2014/05/05 v2.5.5,  2014/06/29 v2.5.6,  2014/10/08 v2.5.7,  2015/02/12 v2.5.8
//  2015/05/03 v2.5.9,  2015/09/24 v2.6.0,  
//  2016/01/17 v2.6.1
// ----------------------------------------------------------------------------
package org.opengts;

import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.text.SimpleDateFormat;

import org.opengts.util.*;

/**
*** This class provides runtime version information to the OpenGTS modules.
**/

public class Version
{

    // ------------------------------------------------------------------------

    // "public" as of 2.3.4-B06
    public  static final String COPYRIGHT = "Copyright 2007-2016, GeoTelematic Solutions, Inc.";

    // ------------------------------------------------------------------------

    // This string is parsed via 'grep' & 'sed' scripts.
    // ONLY the version value specified within the quotes should change.
    private static final String VERSION = "2.6.1";  // 2.6.1-B53, 2.6.0-B87, 2.5.9-B59, 2.5.8-B70

    // This public constant should only be accessed externally by the 'GTSAdmin' application.
    public  static final String GTS_ENTERPRISE_PREFIX = "E";
    public  static final String COMPILED_VERSION = 
        ////  GTS_ENTERPRISE_PREFIX + // "Enterprise"
        VERSION;

    // last compile time
    private static final long COMPILE_TIMESTAMP = CompileTime.COMPILE_TIMESTAMP;

    // ------------------------------------------------------------------------

    // package release (modified by command-line 'sed' script to insert actual epoch time)
    private static final long PACKAGE_TIMESTAMP = 1453075879L;
    public static final String SERVICE_ACCOUNT_ID = "opengts";

    // ------------------------------------------------------------------------
    // ------------------------------------------------------------------------

    /**
    *** Initializes the RTConfig constant 'version' property.
    **/
    public static void initVersionProperty()
    {
        RTProperties constantProps = RTConfig.getRuntimeConstantProperties();
        if (!constantProps.hasProperty(RTKey.VERSION)) {
            constantProps.setProperty(RTKey.VERSION, Version.getVersion());
            //System.out.println("Set Version = " + Version.getVersion());
        } else {
            // already initialized, no need to reinitialize
        }
    };

    // ------------------------------------------------------------------------
    // ------------------------------------------------------------------------

    /**
    *** Returns the copyright notice
    *** @return The copyright notice
    **/
    public static String getCopyright()
    {
        return COPYRIGHT;
    }

    /** 
    *** Returns the compiled version String
    *** @return The version String
    **/
    public static String getVersion()
    {
        return COMPILED_VERSION;
    }

    /** 
    *** Returns the compiled package release timestamp
    *** @return The package release timestamp
    **/
    public static long getPackageTimestamp()
    {
        return PACKAGE_TIMESTAMP;
    }

    /** 
    *** Returns the last compiled timestamp
    *** @return The last compiled timestamp
    **/
    public static long getCompileTimestamp()
    {
        return COMPILE_TIMESTAMP;
    }

    /**
    *** Returns the last compiled time date string
    *** @return The last compiled time date string
    **/
    public static String getCompileTime()
    {
        // (new DateTime(CompileTime.COMPILE_TIMESTAMP))
        return (new DateTime(COMPILE_TIMESTAMP)).toString();
    }

    /** 
    *** Returns the ServiceAccountID
    *** @return The ServiceAccountID
    **/
    public static String getServiceAccountID()
    {
        String sai = RTConfig.getString("ServiceAccount.ID","?");
        return sai.equals(SERVICE_ACCOUNT_ID)? sai : (sai+"("+SERVICE_ACCOUNT_ID+")");
    }

    // ------------------------------------------------------------------------

    /**
    *** Returns a String representation of this Version
    *** @return A String representation of this Version
    **/
    public static String getInfo()
    {
        StringBuffer sb = new StringBuffer();
        sb.append(Version.COPYRIGHT).append("\n");
        sb.append("Version: ").append(Version.getVersion()).append(" [").append(getServiceAccountID()).append("]").append("\n");
        SimpleDateFormat dateFmt = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss zzz", Locale.US);
        if (PACKAGE_TIMESTAMP > 0L) {
            String dfmt = dateFmt.format(new Date(PACKAGE_TIMESTAMP * 1000L));
            sb.append("Package: [" + PACKAGE_TIMESTAMP + "] " + dfmt + "\n");
        }
        if (COMPILE_TIMESTAMP > 0L) {
            String dfmt = dateFmt.format(new Date(COMPILE_TIMESTAMP * 1000L));
            sb.append("Compile: [" + COMPILE_TIMESTAMP + "] " + dfmt + "\n");
        }
        return sb.toString();
    }

    // ------------------------------------------------------------------------

    /**
    *** Main entry point to display the current version
    *** @param argv The command-line args
    **/
    public static void main(String argv[])
    {
        // do not load RTConfig here
        if ((argv.length > 0) && argv[0].equals("-info")) {
            // Version: 2.0.5
            // Compile: [1211915119] 2008/05/27 19:36:38 GMT
            // Package: [1211915119] 2008/05/27 19:36:38 GMT
            System.out.println(Version.getInfo());
        } else
        if ((argv.length > 0) && argv[0].equals("-package")) {
            // "1211915119"
            System.out.println(Version.getPackageTimestamp());
        } else
        if ((argv.length > 0) && argv[0].equals("-compile")) {
            // "1211915119"
            System.out.println(Version.getCompileTimestamp());
        } else {
            // "1.8.3"
            System.out.println(Version.getVersion());
        }
    }

    // ------------------------------------------------------------------------

}
